plugins/inputs/modbus/modbus.go

```diff
@@ -0,0 +1,595 @@
+		}
+
+		r.Fields[i].value = convertDataType(r.Fields[i], bytes)
+	}
+
+	return nil
+}
+
+func convertDataType(t fieldContainer, bytes []byte) interface{} {
+	switch t.DataType {
+	case "UINT16":
+		e16, _ := convertEndianness16(t.ByteOrder, bytes).(uint16)
+		f16 := format16(t.DataType, e16).(uint16)
+		return scale16(t.Scale, f16)
+	case "INT16":
+		e16, _ := convertEndianness16(t.ByteOrder, bytes).(uint16)
+		return format16(t.DataType, e16)
+	case "UINT32":
+		e32, _ := convertEndianness32(t.ByteOrder, bytes).(uint32)
+		f32 := format32(t.DataType, e32).(uint32)
+		return scale32(t.Scale, f32)
+	case "INT32":
+		e32, _ := convertEndianness32(t.ByteOrder, bytes).(uint32)
+		return format32(t.DataType, e32)
+	case "FLOAT32-IEEE":
+		e32, _ := convertEndianness32(t.ByteOrder, bytes).(uint32)
+		return format32(t.DataType, e32)
+	case "FLOAT32":
+		if len(bytes) == 2 {
+			e16, _ := convertEndianness16(t.ByteOrder, bytes).(uint16)
+			f16 := format16(t.DataType, e16).(uint16)
+			return scale16(t.Scale, f16)
+		} else {
+			e32, _ := convertEndianness32(t.ByteOrder, bytes).(uint32)
+			return scale32(t.Scale, e32)
+		}
+	default:
+		return 0
+	}
+}
+
+func convertEndianness16(o string, b []byte) interface{} {
+	switch o {
+	case "AB":
+		return binary.BigEndian.Uint16(b)
+	case "BA":
+		return binary.LittleEndian.Uint16(b)
+	default:
+		return b
+	}
+}
+
+func convertEndianness32(o string, b []byte) interface{} {
+	switch o {
+	case "ABCD":
+		return binary.BigEndian.Uint32(b)
+	case "DCBA":
+		return binary.LittleEndian.Uint32(b)
+	case "BADC":
+		return uint32(binary.LittleEndian.Uint16(b[0:]))<<16 | uint32(binary.LittleEndian.Uint16(b[2:]))
+	case "CDAB":
+		return uint32(binary.BigEndian.Uint16(b[2:]))<<16 | uint32(binary.BigEndian.Uint16(b[0:]))
+	default:
+		return b
+	}
+}
+
+func format16(f string, r uint16) interface{} {
+	switch f {
+	case "UINT16":
+		return r
+	case "INT16":
+		return int16(r)
+	default:
+		return r
+	}
+}
+
+func format32(f string, r uint32) interface{} {
+	switch f {
+	case "UINT32":
+		return r
+	case "INT32":
+		return int32(r)
+	case "FLOAT32-IEEE":
+		return math.Float32frombits(r)
+	default:
+		return r
+	}
+}
+
+func scale16(s string, v uint16) interface{} {
+	if len(s) == 0 {
+		return v
+	}
+
+	n, err := strconv.ParseFloat(s, 32)
+	if err == nil {
+		return float32(v) * float32(n)
+	}
+
+	return 0
+}
+
+func scale32(s string, v uint32) interface{} {
+	if len(s) == 0 {
+		return v
+	}
+
+	n, err := strconv.ParseFloat(s, 32)
+	if err == nil {
+		return float32(v) * float32(n)
+	}
+
+	return 0
+}
+
+// Gather implements the telegraf plugin interface method for data accumulation
+func (m *Modbus) Gather(acc telegraf.Accumulator) error {
+	fields := make(map[string]interface{})
+	tags := make(map[string]string)
+
+	if !m.isConnected {
+		err := connect(m)
+		if err != nil {
+			m.isConnected = false
+			return err
+		}
+	}
+
+	if !m.isInitialized {
+		err := initialization(m)
```

Comment:
This can be part of the `Init() error` function introduced in #5899. no change required

Reply:
✔️ done